src/circuit/gadget/ecc/chip/add.rs

```diff
@@ -0,0 +1,482 @@
+    poly::Rotation,
+};
+use std::collections::HashSet;
+
+#[derive(Clone, Debug)]
+pub struct Config {
+    q_add: Selector,
+    // lambda
+    lambda: Column<Advice>,
+    // x-coordinate of P in P + Q = R
+    pub x_p: Column<Advice>,
+    // y-coordinate of P in P + Q = R
+    pub y_p: Column<Advice>,
+    // x-coordinate of Q or R in P + Q = R
+    pub x_qr: Column<Advice>,
+    // y-coordinate of Q or R in P + Q = R
+    pub y_qr: Column<Advice>,
+    // α = inv0(x_q - x_p)
+    alpha: Column<Advice>,
+    // β = inv0(x_p)
+    beta: Column<Advice>,
+    // γ = inv0(x_q)
+    gamma: Column<Advice>,
+    // δ = inv0(y_p + y_q) if x_q = x_p, 0 otherwise
+    delta: Column<Advice>,
+    // Permutation
+    perm: Permutation,
+}
+
+impl From<&EccConfig> for Config {
+    fn from(ecc_config: &EccConfig) -> Self {
+        Self {
+            q_add: ecc_config.q_add,
+            x_p: ecc_config.advices[0],
+            y_p: ecc_config.advices[1],
+            x_qr: ecc_config.advices[2],
+            y_qr: ecc_config.advices[3],
+            lambda: ecc_config.advices[4],
+            alpha: ecc_config.advices[5],
+            beta: ecc_config.advices[6],
+            gamma: ecc_config.advices[7],
+            delta: ecc_config.advices[8],
+            perm: ecc_config.perm.clone(),
+        }
+    }
+}
+
+impl Config {
+    pub(crate) fn advice_columns(&self) -> HashSet<Column<Advice>> {
+        let mut advice_columns = HashSet::new();
+        advice_columns.insert(self.x_p);
+        advice_columns.insert(self.y_p);
+        advice_columns.insert(self.x_qr);
+        advice_columns.insert(self.y_qr);
+        advice_columns.insert(self.lambda);
+        advice_columns.insert(self.alpha);
+        advice_columns.insert(self.beta);
+        advice_columns.insert(self.gamma);
+        advice_columns.insert(self.delta);
+        advice_columns
+    }
+
+    pub(crate) fn create_gate<F: FieldExt>(&self, meta: &mut ConstraintSystem<F>) {
+        let q_add = meta.query_selector(self.q_add, Rotation::cur());
+        let x_p = meta.query_advice(self.x_p, Rotation::cur());
+        let y_p = meta.query_advice(self.y_p, Rotation::cur());
+        let x_q = meta.query_advice(self.x_qr, Rotation::cur());
+        let y_q = meta.query_advice(self.y_qr, Rotation::cur());
+        let x_r = meta.query_advice(self.x_qr, Rotation::next());
+        let y_r = meta.query_advice(self.y_qr, Rotation::next());
+        let lambda = meta.query_advice(self.lambda, Rotation::cur());
+
+        // α = inv0(x_q - x_p)
+        let alpha = meta.query_advice(self.alpha, Rotation::cur());
+        // β = inv0(x_p)
+        let beta = meta.query_advice(self.beta, Rotation::cur());
+        // γ = inv0(x_q)
+        let gamma = meta.query_advice(self.gamma, Rotation::cur());
+        // δ = inv0(y_p + y_q) if x_q = x_p, 0 otherwise
+        let delta = meta.query_advice(self.delta, Rotation::cur());
+
+        // Useful composite expressions
```

Comment:
I really like this pattern: each of these values are 1 if the if-condition holds, and unconstrained otherwise. We should probably give the pattern a name, although I can't think of one right now.

Reply:
Let's call it the "conditional 1 pattern".